requirements.txt

```diff
@@ -1,51 +1,44 @@
-gunicorn==20.0.4
+gunicorn==20.1.0
 
-Django==3.1.8
-whitenoise==5.1.0
+Django==3.2.6
+whitenoise==5.3.0
 
-django-registration==3.1.2
+django-registration==3.2
 # No update since 2018
 django-cron==0.5.1
 # No update since 2016
 django-npm==1.0.0
-django-watchman==1.1.1
+django-watchman==1.2.0
 django-su==0.9.0
-# No update since 2016
-django-settings-export==1.2.1
-# No update since 2015
-django-macros==0.4.0
-django-mysql==3.7.1
+django-mysql==3.12.0
 # The alternative to this is redis, but we don't use redis yet
 django-constance[database]==2.8.0
+django-webpack-loader==1.1.0
+django-csp==3.7
 
 # graphql
-graphene-django==2.12.1
-django-filter==2.3.0
+graphene-django==2.15.0
+django-filter==2.4.0
 
 # object-level permissions
-rules==2.2
-django-model-utils==4.0.0
+rules==3.0
 
 # These should be okay to update minors
-numpy==1.19.1
-pandas==1.0.5
-pangres==2.1
-
-pyOpenSSL==19.1.0
-protobuf==3.12.2
+numpy==1.21.2
+pandas==1.3.1
+pangres==2.3.1
 
 # These caused problems in the past so test when updating
 djangosaml2==0.19.1
 pysaml2==5.1.0
 
-SQLAlchemy==1.3.18
-psycopg2==2.8.5
-mysqlclient==1.4.6
-google-cloud-bigquery[pandas]==2.4.0
+SQLAlchemy==1.4.22
+psycopg2==2.9.1
+mysqlclient==2.0.3
+google-cloud-bigquery[pandas]==2.24.0
 
-pinax-eventlog==2.0.3
-debugpy==1.2.0
-PyLTI1p3==1.8.6
-django-webpack-loader==0.7.0
+debugpy==1.4.1
 jsonschema==3.2.0
-django-csp==3.6
+pinax-eventlog==5.1.1
+pycryptodome==3.10.1
```

Comment:
@lsloan and @jonespm, I'd like some feedback here. I asked @lsloan to remove `jwtcrypto` in this PR, and I think he took the initiative to see if he could also remove `pycryptodome` -- which worked at the time I guess? However, when I removed `pyOpenSSL`, it seems to have broken the `createkeys` management command (and maybe also `getjwk`). Adding `pycryptodome` back here seems to have fixed the problem.

I know the cryptographic space in Python is a little strange, but it would be nice for maintenance purposes if there was a clearer tie between library marked as a requirement, and library being imported. If `pyOpenSSL` was being used for cryptography (through some indirect way to `Crypto`), we should make note of that. It also seems like `pyOpenSSL` is unmaintained, and the recommendation is to replace it with [`cryptography`](https://cryptography.io/en/latest/). Maybe we ought to replace `pycryptodome` with `cryptography`; that would probably require some more research (see https://cryptography.io/en/latest/faq/#why-use-cryptography).

What do you think would be the appropriate course of action at this point?

Reply:
I think the management command was a nice convenience but the same could probably just be achieved with documentation or with a shell script that just runs something like

```
ssh-keygen -m PEM -t rsa -b 4096 -f jwtRS256.key 
# Don't add a passphrase

openssl rsa -in jwtRS256.key -pubout -outform PEM -out jwtRS256.key.pub
```

or even suggesting people use the IMS KeyGen utility. https://lti-ri.imsglobal.org/keygen/index. In a future version of this code, these keys will be entered into the admin UI so using something like the imsglobal page is really the easiest way. 

I think it's fine if we just make this a separate issue. I'm all for reducing requirement dependencies.

Reply:
Okay, see https://github.com/tl-its-umich-edu/my-learning-analytics/issues/1313